### PR TITLE
[SPARK-21787][SQL] Support for pushing down filters for DateType in native OrcFileFormat

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
## What changes were proposed in this pull request?

This PR support for pushing down filters for DateType in ORC

## How was this patch tested?

Pass the Jenkins with newly add and updated test cases.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
Hi, @cloud-fan, @gatorsmile, @rxin , @sameeragarwal .
I tested this on #18953, this only works with Apache ORC 1.4.0.
This issue is reported saparately. And, to keep #18953 minimally, I didn't include this into that.

**Comment by dongjoon-hyun** [comment]
This will be supported in a new ORCFileFormat after merging #18953.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFilters.scala [hunk 1]
```diff
@@ -82,8 +82,7 @@ private[orc] object OrcFilters {
    * Both CharType and VarcharType are cleaned at AstBuilder.
    */
   private def isSearchableType(dataType: DataType) = dataType match {
-    // TODO: SPARK-21787 Support for pushing down filters for DateType in ORC
-    case BinaryType | DateType => false
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFilters.scala]: This PR supports DateType PPD on new OrcFileFormat.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcFilterSuite.scala [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+ */
+
+package org.apache.spark.sql.execution.datasources.orc
+
+import java.nio.charset.StandardCharsets
+import java.sql.{Date, Timestamp}
+
+import scala.collection.JavaConverters._
+
+import org.apache.orc.storage.ql.io.sarg.{PredicateLeaf, SearchArgument}
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcFilterSuite.scala]: Please note that this suite use Apache ORC `SearchArgument` and `PredicateLeaf` classes.
We cannot share the test code for this suite.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcTest.scala [hunk 3]
```diff
@@ -15,20 +15,21 @@
...
 
 import org.apache.spark.sql._
-import org.apache.spark.sql.hive.test.TestHiveSingleton
 import org.apache.spark.sql.test.SQLTestUtils
 
-private[sql] trait OrcTest extends SQLTestUtils with TestHiveSingleton {
+abstract class OrcTest extends QueryTest with SQLTestUtils {
   import testImplicits._
 
+  protected def format: String
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcTest.scala]: By using full canonical class names, we will explicitly test new OrcFileFormat in `sql/core` and old OrcFileFormat in `sql/hive` without considering SQL ORC switching conf.

**Comment by dongjoon-hyun** [comment]
Hi, @cloud-fan .
This is the first followup after #19651 .
Could you review this PR, too?

**Comment by cloud-fan** [comment]
maybe have a PR to move the tests first?

**Comment by dongjoon-hyun** [comment]
@cloud-fan .

For the other part, I'm restructuring to remove redundancy in the same manner of `InMemoryCatalogedDDLSuite`, `HiveCatalogedDDLSuite`, and `DDLSuite`.

This one cannot be the part of that restructuring, because the test function signatures are different due to `PredicateLeaf` and `SearchArgument`.

Could you review this first?

**Comment by dongjoon-hyun** [comment]
BTW, @cloud-fan . Do you mean literally `move`? I'm wondering if I'm thinking in a different way.

**Comment by dongjoon-hyun** [comment]
Hi, @cloud-fan .
Could you review this PR again?

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by dongjoon-hyun** [comment]
Thank you so much, @cloud-fan !

**Pull request closed without merging**